34-chinese-pangolin/technical_specification.md

```diff
@@ -0,0 +1,83 @@
+# File Deletion and Outbox Cache Strategy (Chinese Pangolin)
+**Epic Type:** Implementation Epic
+
+## Scope
+### Outline:
+
+This Epic adds a new service, which provides a deletion endpoint for files.
+Calling this enpoint internally publishes an event.
+All services subscribing to this event will then delete this file from their databases and from S3 buckets they have write access to.
+
+### Included/Required:
+
+This Epic includes:
+
+- The creation of a new service, the **Purge Controller Service (PCS)**, which provides a RESTfull HTTP DELETE enpoint /files/{file_id}.
+- Upon receiving and validating the deletion request, it publishes an event
+- The following services subscribe to the new event and do the following after consuming:
+    - IFRS: Deletes the file from permanent storage (IRS would be better, but IRS does not have the connection between outward-facing file_id and S3 file name)
+    - IFRS: Send API Call to EKSS to delete the secret
+    - IFRS: Delete file from outbox
+    - IFRS: Delete file entry from its database
+    - DCS: Delete file entry from its database
+    - Both services: Publish confirmation event afterwards
+- The EKSS provides a new RESTful HTTP DELETE endpoint /sercets/{secret_id} that deletes the corresponding secret from vault.
+
+These parts of the epic are described in detail below.
+
+### Optional: Caching Strategy for DCS/Outbox
+
+- Add a "last_accessed" field for each entry in the DCS files database
+- Each time, this file is sucessfully accessed via a download request, update the "last_accessed" field for this files entry
+- Provide Script that looks through all files currently in the Outbox. If their "last_accessed" time is longer ago then a certain threshold, delete that file.
+
+### Not included:
+
+Any functionality concerning caching/cleanup of the upload path.
+
+- Delete file entries from UCS Database
+- Delete files from Inbox
+- Delete files from Staging
+
+## API Definitions:
+
+### RESTful/Synchronous:
+
+The PCS should have a single endpoint that can be accessed to delete files, conforming to the GDPRs right to be forgotten.
+
+- PCS: DELETE /files/{file_id}
+    - auth header: internal access token
+    - Response: 204 'No Content'
+    - If the validation of the auth header fails: 403 'Forbidden'
+
+This will internally send out a deletion event (See below.)
+
+The EKSS will get a corresponding endpoint to delete a file secret.
+
+- EKSS: DELETE /secrets/{secret_id}:
+    - no header, no body
```

Comment:
This means this call is unauthenticated and guarded by only being available internally?
Probably missed that part of the discussion, so what's the rationale here?

Reply:
That's pretty much the rational right now. We might change it later.